### PR TITLE
Docs: DO-2673 add docs page about persistence and collaboration

### PR DESCRIPTION
**Pull request by krzysztof-causalens**
## Motivation and Context

The recently added collaboration/persistence feature wasn't documented outside of the changelog so adding a page to dive deeper into how it works.

Linked from the sidebar and in a tip in the interactivity section.

## PR Checklist:

- [x] I have implemented all requirements? (see JIRA, project documentation).
- [x] I am not affecting someone else's work, If I am, they are included as a reviewer.
- [x] I have added relevant tests (unit, integration or regression).
- [ ] I have added comments to all the bits that are hard to follow.
- [ ] I have added/updated Documentation.
- [ ] I have updated the appropriate changelog with a line for my changes.

### CODE REVIEW

File: docs/advanced/persistence-and-collaboration.mdx [hunk 1]
```diff
@@ -0,0 +1,102 @@
+---
+title: Persistence and Collaboration
+---
+
+By default `Variable` state is purely client-side - by default its source of truth is the browser memory.
```
Comment (patricia-causalens) [inline docs/advanced/persistence-and-collaboration.mdx:5]: I think it could be nice to add a small sentence or two just about what this page is about. For if someone is flicking through our docs and wants to have a little summary before committing to reading the whole page.

File: docs/advanced/persistence-and-collaboration.mdx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+user and act on their behalf, e.g. `write` would update the value stored only for that particular user.
+
+:::
+
+### `backend`
+
+You can select a particular backend implemention to be used for storage. By default `BackendStore` uses the
+`dara.core.persistence.InMemoryBackend` implementation which simply stores the data in-memory.
+
+Alternatively, you can use the `FileBackend` which uses a JSON file for storage. This enables persisting
```
Comment (patricia-causalens) [inline docs/advanced/persistence-and-collaboration.mdx:95]: Could we have a little example using the FileBackend? I think initially that will be a common one to use

File: packages/dara-core/dara/core/persistence.py [hunk 3]
```diff
@@ -281,7 +281,8 @@ async def write(self, value: Any, notify=True):
         """
         Persist a value to the store.
 
-        If scope='user', the value is written for the current user.
+        If scope='user', the value is written for the current user so the method can only
+        be used in authenticated contexts.
```
Comment (patricia-causalens) [inline packages/dara-core/dara/core/persistence.py:285]: Do we have docs to point to here for authenticated contexts? I think that is something that will be confusing to people

**Review by patricia-causalens — CHANGES_REQUESTED** [review]
Looks good, just a few small comments

File: packages/dara-core/dara/core/persistence.py (hunk 3 above)
Comment (krzysztof-causalens) [reply packages/dara-core/dara/core/persistence.py:285]: We don't have a dedicated page but it is mentioned a couple times. Perhaps deserves a separate authentication related page (I think there was an issue raised on GH asking for that) but I think that's a separate effort, will try to remember to link it here once we get around to it

File: docs/advanced/persistence-and-collaboration.mdx (hunk 2 above)
Comment (krzysztof-causalens) [reply docs/advanced/persistence-and-collaboration.mdx:95]: Good point, will add an example

File: packages/dara-core/dara/core/persistence.py (hunk 3 above)
Comment (patricia-causalens) [reply packages/dara-core/dara/core/persistence.py:285]: Fair, could even be on the FAQ depending on how much we want to say there

File: docs/advanced/persistence-and-collaboration.mdx (hunk 1 above)
Comment (krzysztof-causalens) [reply docs/advanced/persistence-and-collaboration.mdx:7]: I mean the sentences here were meant to be a summary, it's a bit lengthy but it's tricky to explain. How does the current version sound?

Comment (krzysztof-causalens) [reply docs/advanced/persistence-and-collaboration.mdx:7]: I realise I made it even longer but I wanted to paint a picture of why this is useful

Comment (patricia-causalens) [reply docs/advanced/persistence-and-collaboration.mdx:7]: Looks better! How about, instead of:
In order to make the aforementioned default behaviour more customizable to suit different needs, `Variable` accepts a `store` property. It can be used to customize the source of truth for a given `Variable` and enables
persistence and collaboration.

Something like:

To address these limitations, `Variable` accepts a `store` property. This allows for customization of the Variable's source of truth, thereby enabling features like data persistence and collaborative interactions across users.

Comment (krzysztof-causalens) [reply docs/advanced/persistence-and-collaboration.mdx:7]: I wouldn't call it a limitation, it's just how it works 😅 Will try to find a better word, but sure looks better

**Review by patricia-causalens — APPROVED** [review]
LGTM!

File: docs/advanced/persistence-and-collaboration.mdx [hunk 4]
```diff
@@ -0,0 +1,115 @@
+---
+title: Persistence and Collaboration
+---
+
+## Introduction
+
+By default `Variable` state is purely client-side - by default its source of truth is the browser memory.
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:7]: By default `Variable` state is purely client-side -> By default a `Variable`'s state is purely client-side 

Also this sentence feels a bit awkward maybe just putting a connector word:

.. is purely client-side, meaning its source of truth is the browser memory.

File: docs/advanced/persistence-and-collaboration.mdx [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+It also means that refreshing the browser page resets the `Variable` state, and that each user has their own 'copy'
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:8]: 'copy' -> "copy"

File: docs/advanced/persistence-and-collaboration.mdx [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+of the data. To customize the behaviour, `Variable` accepts a store property. This allow changing the `Variable`'s source of truth,
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:9]: behaviour  -> behavior 
allow -> allows

Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:9]: we seem to be switching between referring to `Variable` without any articles and referring to it with articles ie referring to "a" or "the" `Variable` or just referring to the object as the class itself. I don't mind either way but let's probably stick with one

File: docs/advanced/persistence-and-collaboration.mdx [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+The rest of the page goes in-depth into available stores and how they can be used in your application.
+
+
+## `dara.core.persistence.BackendStore`
+
+This type of store changes the source of truth for a given `Variable` to live on the server, rather than on the client.
+This has two main benefits:
+- the data can be persistent across page refreshes or even across server restarts
+- since the data lives server-side, we can 'sync' it across clients to enable collaboration
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:20]: 'sync' -> "sync"

File: docs/advanced/persistence-and-collaboration.mdx [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+To use a `BackendStore`, simply create a store instance and attach it to a `Variable`:
+
+```python
+from dara.core import Variable
+from dara.core.persistence import BackendStore
+from dara.components import Input
+
+# Tag the variable to be 'collaborative' and have a shared server state
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:29]: 'collaborative' -> "collaborative"

File: docs/advanced/persistence-and-collaboration.mdx [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+collab_variable = Variable(default=1, store=BackendStore())
+
+# Input value will automatically be kept in sync 'live' for all users
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:32]: 'live' -> "live"

File: docs/advanced/persistence-and-collaboration.mdx [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+from dara.core.persistence import BackendStore
+from dara.components import Input
+
+store = BackendStore()
+collab_variable = Variable(default='default value', store=store)
+
+# get current value in the store
+value = await store.read()
+# store can also be accessed from the variable
+value = await variable.store.read()
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:55]: I am guessing these print the same thing?

File: docs/advanced/persistence-and-collaboration.mdx [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+await store.write('new value')
+
+# delete the existing value from the store
+await store.delete()
+
+# get all value from the store - see `scope` section below for details
+await store.get_all()
+```
+
+The behaviour of the store can be customized by passing arguments to the `BackendStore`.
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:67]: behaviour -> behavior

File: docs/advanced/persistence-and-collaboration.mdx [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+### `scope`
+
+Scope controls whether the data in a `BackendStore` is shared among all users (`scope='global'`) or separate and specific to a
+given user (`scope='user'`).
+
+When using `global` scope, the state is shared and synced among the users so a change being made to an attached `Variable`
+in one client will immediately be propagated to other users. In this mode, the `get_all` API method will return a dictionary
+of the shape `{'global': 'value'}`.
+
+When using `user` scope, each user has their own copy of the date. A change made to an attached `Variable` in one client will
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:78]: date -> data

File: docs/advanced/persistence-and-collaboration.mdx [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+only be propagated to other sessions open **for that particular user**, e.g. among all currently open tabs for that user.
+In this mode, the `get_all` API returns a dictionary with user values keyed by the user identifiers, e.g. `{'user1': 'value1', 'user2': 'value2'}`.
+
+:::note
+
+In `user` scope, the API methods `read`, `write` and `delete` can only be used in an authenticated context, i.e. within
```
Comment (claudiawestby) [inline docs/advanced/persistence-and-collaboration.mdx:84]: I might even elevate this to warning, this has been something that has tripped up even @patricia-causalens  and I a few times :D

**Review by claudiawestby — APPROVED** [review]
Very nice docs! just left some minor comments

File: docs/advanced/persistence-and-collaboration.mdx (hunk 10 above)
Comment (krzysztof-causalens) [reply docs/advanced/persistence-and-collaboration.mdx:55]: Yeah that's the idea, I'm just trying to show that you can access it from the variable so you don't need to keep around a reference to the store itself.
